File: src/Solutions/Euler084.hs
module Solutions.Euler084 where
import Utils.Array(modifyArray)
import qualified Utils.Queue as Q
import Control.Monad
import Control.Monad.ST
import Control.Monad.Trans
import Control.Monad.Trans.State.Strict
import Data.Array.ST
import Data.List(sortBy)
import Data.STRef
import System.Random
import Text.Printf


newtype Square = Square Int deriving (Eq, Ord, Ix)
instance Show Square where show (Square n) = printf "%02d" n

squareGO = Square 0
squareR1 = Square 5
squareJAIL = Square 10
squareC1 = Square 11
squareE3 = Square 24
squareG2J = Square 30
squareH2 = Square 39
squaresCC = [Square 2, Square 17, Square 33]
squaresCH = [Square 7, Square 22, Square 36]
nextR (Square n) | n > 5 && n < 15 = Square 15
                 | n > 15 && n < 25 = Square 25
                 | n > 25 && n < 35 = Square 35
                 | otherwise = Square 5
nextU (Square n) | n > 12 && n < 28 = Square 28
                 | otherwise = Square 12

applyStep :: Int -> Square -> Square
applyStep step (Square n) =
    let newSquare = Square $ (n + step) `mod` 40
    in if newSquare == squareG2J then squareJAIL else newSquare


type VisitedCounter s = STUArray s Square Int

getAnswer :: [(Square, Int)] -> String
getAnswer assocs =
    let [(max1, _), (max2, _), (max3, _)] = take 3 $ sortBy (\(_, a) (_, b) -> compare b a) assocs
    in show max1 ++ show max2 ++ show max3


data Card = Go | Jail | C1 | E3 | H2 | R1 | NextR | NextU | Back3 | Null
communityCards = [Go, Jail] ++ replicate 14 Null
chanceCards = [Go, Jail, C1, E3, H2, R1, NextR, NextR, NextU, Back3] ++ replicate 6 Null
newtype Deck = Deck (Q.Queue Card)
mkDeck :: [Card] -> Deck
mkDeck cards = Deck (Q.fromList cards)

takeCard :: Deck -> (Card, Deck)
takeCard (Deck q) =
    let (Just card, q') = Q.dequeue q
    in (card, Deck $ Q.enqueue card q')


data GameState s = GameState
    { position :: Square
    , visitedCounter :: VisitedCounter s
    , doubles :: Int
    , communityDeck :: Deck
    , chanceDeck :: Deck
    , gen :: StdGen
    }

initGameState :: StdGen -> VisitedCounter s -> GameState s
initGameState g counter =
    let (communityCardsShuffled, g') = shuffle communityCards g
        (chanceCardsShuffled, g'') = shuffle chanceCards g'
    in GameState squareGO counter 0
                 (mkDeck communityCardsShuffled)
                 (mkDeck chanceCardsShuffled)
                 g''


type SimulationState s a = StateT (GameState s) (ST s) a
takeCommunityCard :: SimulationState s Card
takeCommunityCard = get >>= \state ->
    let (card, newCommunityDeck) = takeCard (communityDeck state)
    in put (state { communityDeck = newCommunityDeck }) >> return card

applyCommunityCard :: Square -> SimulationState s ()
applyCommunityCard sq = takeCommunityCard >>= \card -> goToSquare $ case card of
    Go -> squareGO
    Jail -> squareJAIL
    Null -> sq

takeChanceCard :: SimulationState s Card
takeChanceCard = get >>= \state ->
    let (card, newChanceDeck) = takeCard (chanceDeck state)
    in put (state { chanceDeck = newChanceDeck }) >> return card

applyChanceCard :: Square -> SimulationState s ()
applyChanceCard sq = takeChanceCard >>= \card -> goToSquare $ case card of
    Go -> squareGO
    Jail -> squareJAIL
    C1 -> squareC1
    E3 -> squareE3
    H2 -> squareH2
    R1 -> squareR1
    NextR -> nextR sq
    NextU -> nextU sq
    Back3 -> applyStep (-3) sq
    Null -> sq

rollDices :: SimulationState s Int
rollDices = get >>= \state ->
    let (roll1, gen') = randomR (1, 4) $ gen state
        (roll2, gen'') = randomR (1, 4) gen'
        doubles' = if roll1 == roll2 then doubles state + 1 else 0
    in put (state { gen = gen'', doubles = doubles' }) >> return (roll1 + roll2)

getNextPosition :: SimulationState s Square
getNextPosition = do
    step <- rollDices
    dbls <- gets doubles
    if dbls == 3
        then return squareJAIL
        else applyStep step <$> gets position

goToSquare :: Square -> SimulationState s ()
goToSquare sq = do
    state <- get
    lift $ modifyArray (visitedCounter state) sq succ
    put state { position = sq }

movePawn :: Square -> SimulationState s ()
movePawn newPosition
    | newPosition `elem` squaresCC = applyCommunityCard newPosition
    | newPosition `elem` squaresCH = applyChanceCard newPosition
    | otherwise = goToSquare newPosition

simulateGame :: Int -> SimulationState s ()
simulateGame moves = replicateM_ moves (getNextPosition >>= movePawn)


makeArray :: [a] -> ST s (STArray s Int a)
makeArray list = newListArray (1, length list) list

shuffle :: [a] -> StdGen -> ([a], StdGen)
shuffle list g = runST $ do
    gen <- newSTRef g
    arr <- makeArray list
    forM_ [length list..2] $ \i -> do
        (j, g') <- randomR (1, i) <$> readSTRef gen
        writeSTRef gen g'
        tmp <- readArray arr i
        readArray arr j >>= writeArray arr i
        writeArray arr j tmp
    shuffledList <- getElems arr
    g' <- readSTRef gen
    return (shuffledList, g')

totalMoves = 2000000 :: Int
evalGame :: StdGen -> String
evalGame g = runST $ do
    arr <- newArray (squareGO, squareH2) 0 :: ST s (VisitedCounter s)
    evalStateT (simulateGame totalMoves) (initGameState g arr)
    getAnswer <$> getAssocs arr


euler084 :: IO String
euler084 = evalGame <$> newStdGen

File: rust/src/solutions/euler084.rs
use rand::{
    distributions::{Distribution, Uniform},
    rngs::ThreadRng,
    seq::SliceRandom,
};
use std::ops::{Add, AddAssign};

#[derive(Copy, Clone, PartialEq, Eq)]
struct Square(usize);

const SQUARE_GO: Square = Square(0);
const SQUARE_R1: Square = Square(5);
const SQUARE_JAIL: Square = Square(10);
const SQUARE_C1: Square = Square(11);
const SQUARE_E3: Square = Square(24);
const SQUARE_G2J: Square = Square(30);
const SQUARE_H2: Square = Square(39);
const SQUARES_CC: [Square; 3] = [Square(2), Square(17), Square(33)];
const SQUARES_CH: [Square; 3] = [Square(7), Square(22), Square(36)];

fn next_r(&Square(n): &Square) -> Square {
    Square(match n {
        _ if n > 5 && n < 15 => 15,
        _ if n > 15 && n < 25 => 25,
        _ if n > 25 && n < 35 => 35,
        _ => 5,
    })
}

fn next_u(&Square(n): &Square) -> Square {
    Square(match n {
        _ if n > 12 && n < 28 => 28,
        _ => 12,
    })
}

impl Add<usize> for Square {
    type Output = Square;
    fn add(self, rhs: usize) -> Self::Output {
        let Square(n) = self;
        Square((n + rhs) % 40)
    }
}

impl AddAssign<usize> for Square {
    fn add_assign(&mut self, rhs: usize) {
        *self = *self + rhs;
    }
}

#[derive(Copy, Clone, PartialEq)]
enum Card {
    Go,
    Jail,
    C1,
    E3,
    H2,
    R1,
    NextR,
    NextU,
    Back3,
    Null,
}

struct GameState {
    position: Square,
    visited_counter: [usize; 40],
    doubles: u8,
    community_deck: Vec<Card>,
    chance_deck: Vec<Card>,
    gen: ThreadRng,
    dice: Uniform<usize>,
}

impl GameState {
    pub fn new() -> GameState {
        let mut gen = rand::thread_rng();
        let mut community_deck = vec![Card::Go, Card::Jail];
        community_deck.extend(&[Card::Null; 14]);
        community_deck.shuffle(&mut gen);
        let mut chance_deck = vec![
            Card::Go,
            Card::Jail,
            Card::C1,
            Card::E3,
            Card::H2,
            Card::R1,
            Card::NextR,
            Card::NextR,
            Card::NextU,
            Card::Back3,
        ];
        chance_deck.extend(&[Card::Null; 6]);
        chance_deck.shuffle(&mut gen);
        GameState {
            position: SQUARE_GO,
            visited_counter: [0; 40],
            doubles: 0,
            community_deck,
            chance_deck,
            gen,
            dice: Uniform::from(1..5),
        }
    }

    fn roll_dice(&mut self) -> usize {
        self.dice.sample(&mut self.gen)
    }

    fn go_to_square(&mut self, &square: &Square) {
        let Square(n) = match square {
            SQUARE_G2J => SQUARE_JAIL,
            _ => square,
        };
        self.visited_counter[n] += 1;
        self.position = Square(n);
    }

    fn interpret_card(&mut self, &card: &Card) {
        match card {
            Card::Go => self.go_to_square(&SQUARE_GO),
            Card::Jail => self.go_to_square(&SQUARE_JAIL),
            Card::C1 => self.go_to_square(&SQUARE_C1),
            Card::E3 => self.go_to_square(&SQUARE_E3),
            Card::H2 => self.go_to_square(&SQUARE_H2),
            Card::R1 => self.go_to_square(&SQUARE_R1),
            Card::NextR => self.go_to_square(&next_r(&self.position)),
            Card::NextU => self.go_to_square(&next_u(&self.position)),
            Card::Back3 => self.go_to_square(&(self.position + 37)), // -3 == +37 mod 40
            Card::Null => self.go_to_square(&self.position.clone()),
        }
    }

    fn draw_community_card(&mut self) {
        self.interpret_card(&self.community_deck[0].clone());
        self.community_deck.rotate_left(1);
    }

    fn draw_chance_card(&mut self) -> () {
        self.interpret_card(&self.chance_deck[0].clone());
        self.chance_deck.rotate_left(1);
    }

    fn move_pawn(&mut self, new_position: Square) {
        match new_position {
            _ if SQUARES_CC.contains(&new_position) => {
                self.position = new_position;
                self.draw_community_card();
            }
            _ if SQUARES_CH.contains(&new_position) => {
                self.position = new_position;
                self.draw_chance_card();
            }
            _ => self.go_to_square(&new_position),
        }
    }

    fn get_next_position(&mut self) -> Square {
        let roll1: usize = self.roll_dice();
        let roll2: usize = self.roll_dice();
        self.doubles = if roll1 == roll2 { self.doubles + 1 } else { 0 };
        if self.doubles == 3 {
            self.doubles = 0;
            SQUARE_JAIL
        } else {
            self.position + (roll1 + roll2)
        }
    }

    pub fn simulate_game(&mut self, steps: usize) {
        for _ in 0..steps {
            let next_pos = self.get_next_position();
            self.move_pawn(next_pos);
        }
    }

    pub fn get_answer(&self) -> String {
        let mut answers: Vec<(usize, &usize)> =
            (0usize..).zip(self.visited_counter.iter()).collect();
        answers.sort_by_key(|n| n.1);
        format!(
            "{:02}{:02}{:02}",
            answers[39].0, answers[38].0, answers[37].0
        )
    }
}

pub fn euler084() -> String {
    let mut state = GameState::new();
    state.simulate_game(2000000);
    state.get_answer()
}
